### PR TITLE
MAP estimator for bayes

### PR DESCRIPTION
**Pull request by msultan**
- [x] Implement feature / fix bug
 - [ ] Update changelog
 - [ ] Test added

MAP estimator based upon issue raised in #752 . I made the choice of using the mean of the posterior logliklihoods though an argument could be made for the median of the distribution too based upon 
https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation

### CODE REVIEW

**Comment by mpharrigan** [comment]
Why did you remove my lovely checkbox for "test added" :)

**Comment by msultan** [comment]
@mpharrigan and here i thought I was being all sneaky about this.
this is ready for code review btw

**Comment by cxhernandez** [comment]
It failed on Travis and pass Appveyor...? There's something new...

File: msmbuilder/msm/bayesmsm.py [hunk 1]
```diff
@@ -235,6 +247,15 @@ def _fit_reversible(self, countsmat):
         # we round chain_length UP, so we might have generated a few extra
         # samples.
         result = result[-self.n_samples:]
+
+        self.all_logliklihoods_ = [_logprob_T(i,Z) for i in result]
```
Comment (cxhernandez) [inline msmbuilder/msm/bayesmsm.py]: `all_loglikelihoods_`?

Comment (msultan) [inline msmbuilder/msm/bayesmsm.py]: damn it

**Comment by msultan** [comment]
Anyone have strong opinions on whether we should report the mean(as it is now) or the median reversible estimator as the MAP estimate?

**Comment by msultan** [comment]
Interestingly enough, the start of each iteration on each of my cpu cores seems to give the best mdl. That is not too surprising since i think its just returning the unperturbed tmat back 

image

**Comment by msultan** [comment]
Also if someone has any papers about how to change this sampler so that it doesnt require a reversible counts matrix, I am all ears. Currently, it adds the transpose to the counts matrix which I think is major flaw since it gets rid of a lot of interesting structure within our data. It technically shouldn't matter when we have a lot of transitions but that is rarely the case for larger systems of interest.

**Comment by msultan** [comment]
image
Here is the histogram for one of my datasets . Its pretty gaussian-y so mean/median should both be fine.

**Comment by msultan** [comment]
@cxhernandez @mpharrigan anyone wanna read through the wiki page and make sure this is implemented correctly?

**Comment by rmcgibbo** [comment]
this code doesn't make any sense to me. why are you taking the mean of the log posterior?

**Comment by msultan** [comment]
@rmcgibbo I wanted the mode of the posterior distribution to report a single best model from the set of all mdls reported by the BayesianMSM. However, that would have required some binning. I thought it might be better to take the mean of the distribution under the assumption that it would be close to the mode given that the distribution is gaussian.

**Comment by rmcgibbo** [comment]
the "map" is the _maximum_ of the posterior density.

**Comment by msultan** [comment]
@rmcgibbo yea, i think i misunderstood that. I will delete this PR. what would be a good way to get the best representative model out from the Bayesian MSM.

**Comment by msultan** [comment]
btw, thanks a lot for the catch @rmcgibbo . I completely misunderstood, what distribution I was supposed to use.

**Pull request closed without merging**